src/devices/sound/upd933.cpp

```diff
@@ -113,40 +123,85 @@ void upd933_device::device_clock_changed()
 	m_stream->set_sample_rate(clock() / CLOCKS_PER_SAMPLE);
 }
 
+/**************************************************************************/
+int upd933_device::rq_r()
+{
+	if (!machine().side_effects_disabled())
+		m_stream->update();
+
+	return m_irq_state;
+}
+
 /**************************************************************************/
 void upd933_device::cs_w(int state)
 {
 	m_stream->update();
 
 	if (!m_cs && state)
-		check_irq();
+		update_pending_irq();
 	m_cs = state;
+	update_irq();
+}
+
+/**************************************************************************/
+void upd933_device::id_w(int state)
+{
+	m_stream->update();
+
+	m_id = state;
+	update_irq();
 }
 
 /**************************************************************************/
-void upd933_device::check_irq()
+u8 upd933_device::irq_data() const
 {
+	// TODO: do these have the correct priority?
 	for (int i = 0; i < 8; i++)
 	{
-		if (m_dca[i].m_irq)
-		{
-			m_irq_data = 1 | (i << 1);
-			break;
-		}
-		if (m_dcw[i].m_irq)
-		{
-			m_irq_data = 2 | (i << 2);
-			break;
-		}
 		if (m_dco[i].m_irq)
-		{
-			m_irq_data = 4 | (i << 3);
-			break;
-		}
+			return 4 | (i << 3);
+	}
+	for (int i = 0; i < 8; i++)
+	{
+		if (m_dcw[i].m_irq)
+			return 2 | (i << 2);
 	}
+	for (int i = 0; i < 8; i++)
+	{
+		if (m_dca[i].m_irq)
+			return 1 | (i << 1);
+	}
+	return 0;
+}
 
-	if (m_irq_data)
-		m_irq_cb(m_irq_state = 1);
+/**************************************************************************/
+void upd933_device::update_pending_irq()
+{
+	m_irq_pending = 0;
+	bool env_active = false;
+	unsigned new_time = UINT_MAX;
+
+	for (int i = 0; i < 8; i++)
+	{
+		env_active |= (m_dca[i].calc_timeout(new_time)
+		           |   m_dco[i].calc_timeout(new_time)
+		           |   m_dcw[i].calc_timeout(new_time));
+	}
+
+	if (!new_time)
+		m_irq_pending = 1;
+	else if (env_active)
+		m_irq_timer->adjust(clocks_to_attotime((u64)new_time * CLOCKS_PER_SAMPLE));
+}
+
+/**************************************************************************/
+void upd933_device::update_irq()
+{
+	if (m_irq_state != (m_cs & m_id & m_irq_pending))
+	{
+		m_irq_state ^= 1;
+		m_irq_cb(m_irq_state);
+	}
```

Comment:
If someone puts an unexpected value into `m_irq_state` (e.g. with the debugger or a Lua script, since they can access stuff registered for save states) this is going to behave *really* strangely.